Sources/SKCore/TaskScheduler.swift

```diff
@@ -0,0 +1,459 @@
+  ///
+  /// `nonisolated(unsafe)` is fine because it will never get modified after being set in the initializer.
+  nonisolated(unsafe) private(set) var resultTask: Task<Void, Never>! = nil
+
+  /// After `execute` is called, the `suspendableTask` is a task that performs the computation defined by `description`.
+  /// The `resultTask` effectively waits for this task to be set (by watching for new values produced by
+  /// `suspendableTaskCreatedContinuation`) and awaits its result. The task can terminate with two different statuses:
+  ///  - `terminated`: The task has finished executing and the `resultTask` is done
+  ///  - `suspended`: The `suspendableTask` was cancelled by calling `QueuedTask.suspend()`. In this case the
+  ///    `TaskScheduler` is expected to call `execute` again, which will produce a new `suspendableTask`. `resultTask`
+  ///    then awaits this new `suspendableTask` and awaits its result.
+  private var suspendableTask: Task<SuspendableTaskFinishStatus, Never>?
+
+  /// Every time `execute` gets called, a new task is placed in this continuation. See comment on `suspendableTask`.
+  private let suspendableTaskCreatedContinuation: AsyncStream<Task<SuspendableTaskFinishStatus, Never>>.Continuation
+
+  /// Placing a new value in this continuation will cause `resultTask` to query its priority and set
+  /// `QueuedTask.priority`.
+  private let updatePriorityContinuation: AsyncStream<Void>.Continuation
+
+  nonisolated(unsafe) private var _priority: AtomicUInt8
+
+  /// The latest known priority of the task.
+  ///
+  /// This starts off as the priority with which the task is being created. If higher priority tasks start depending on
+  /// it, the priority may get elevated.
+  nonisolated var priority: TaskPriority {
+    get {
+      TaskPriority(rawValue: _priority.value)
+    }
+    set {
+      _priority.value = newValue.rawValue
+    }
+  }
+
+  /// Whether `suspendDependency` has been called on this `QueuedTask`.
+  ///
+  /// Gets reset every time `suspendableTask` finishes.
+  nonisolated(unsafe) private var suspended: AtomicBool = .init(initialValue: false)
+
+  init(priority: TaskPriority? = nil, description: TaskDescription) async {
+    self._priority = .init(initialValue: priority?.rawValue ?? Task.currentPriority.rawValue)
+    self.description = description
+
+    var updatePriorityContinuation: AsyncStream<Void>.Continuation!
+    let updatePriorityStream = AsyncStream {
+      updatePriorityContinuation = $0
+    }
+    self.updatePriorityContinuation = updatePriorityContinuation
+
+    var suspendableTaskCreatedContinuation: AsyncStream<Task<SuspendableTaskFinishStatus, Never>>.Continuation!
+    let suspendableTaskCreatedStream = AsyncStream {
+      suspendableTaskCreatedContinuation = $0
+    }
+    self.suspendableTaskCreatedContinuation = suspendableTaskCreatedContinuation
+
+    self.resultTask = Task.detached(priority: priority) {
+      await withTaskGroup(of: Void.self) { taskGroup in
```

Comment:
I don’t think that works because I need to cancel the first task in the task group (the one that is updating `self.priority`) when the second task in the task group (the one that runs tasks from `executionTaskCreatedStream`) finishes.